File: scala/src/main/scala/yuzu/dataid.scala
package com.github.jmccrae.yuzu

import com.github.jmccrae.yuzu.YuzuSettings._
import com.hp.hpl.jena.datatypes.xsd.XSDDateTime
import com.hp.hpl.jena.graph.NodeFactory
import com.hp.hpl.jena.rdf.model.{Model, ModelFactory}
import com.hp.hpl.jena.vocabulary._

object DataID {

  def get : Model = {
    val model = ModelFactory.createDefaultModel()
    val backend = new TripleBackend(DB_FILE)

    val dataid = model.createResource(BASE_NAME + METADATA_PATH)

    dataid.addProperty(RDF.`type`, model.createResource(DCAT + "Dataset"))
    dataid.addProperty(RDF.`type`, model.createResource(VOID + "Dataset"))

    dataid.addProperty(
      model.createProperty(DCAT + "title"),
      DISPLAY_NAME, LANG)

    dataid.addProperty(
      RDFS.label,
      DISPLAY_NAME, LANG)

    dataid.addProperty(
      model.createProperty(DCAT + "landingPage"),
      model.createResource(BASE_NAME))
    
    backend.listResources(0, 1)._2.headOption match {
      case Some(SearchResult(link, _)) =>
        dataid.addProperty(
          model.createProperty(VOID + "exampleResource"),
          model.createResource(BASE_NAME.dropRight(1) + link))
      case None =>
    }

    dataid.addProperty(
      DC_11.language,
      LANG)

    dataid.addProperty(
      model.createProperty(VOID + "rootResource"),
      model.createResource(BASE_NAME.dropRight(1) + LIST_PATH))

    ONTOLOGY match {
      case Some(o) =>
        dataid.addProperty(
          model.createProperty(DATAID + "ontologyLocation"),
          model.createResource(BASE_NAME + o))
      case None =>
    }

    ISSUE_DATE match {
      case Some(id) =>
        dataid.addProperty(
          DCTerms.issued,
          id, NodeFactory.getType(XSD.date.getURI()))
      case None =>
    }

    VERSION_INFO match {
      case Some(vi) =>
        dataid.addProperty(
          model.createProperty(DATAID + "versionInfo"),
          vi)
      case None =>
    }

    DESCRIPTION match {
      case Some(d) => 
        dataid.addProperty(
          DC_11.description,
          d, LANG)
      case None =>
    }

    dataid.addProperty(
      model.createProperty(ODRL + "license"),
      model.createResource(BASE_NAME.dropRight(1) + LICENSE_PATH))

    LICENSE match {
      case Some(l) =>
        dataid.addProperty(
          DC_11.rights,
          model.createResource(l))
      case None => 
    }

    for(keyword <- KEYWORDS) {
      dataid.addProperty(
        model.createProperty(DCAT + "keyword"),
        keyword)
    }

    PUBLISHER_NAME match {
      case Some(pn) =>
        val publisher = model.createResource(BASE_NAME + METADATA_PATH + "#Publisher")
        publisher.addProperty(RDF.`type`, model.createResource(FOAF + "Agent"))
        publisher.addProperty(RDF.`type`, model.createResource(PROV + "Agent"))
        dataid.addProperty(
          DC_11.publisher,
          publisher)
        publisher.addProperty(
          model.createProperty(FOAF + "name"),
          pn)
        PUBLISHER_EMAIL match {
          case Some(pe) =>
            publisher.addProperty(
              model.createProperty(FOAF + "mbox"),
              pe)
          case None =>
        }
      case None => 
    }

    for(((cn, ce), id) <- (CREATOR_NAMES zip CREATOR_EMAILS).zipWithIndex) {
      val creator = model.createResource(BASE_NAME + METADATA_PATH + "#Creator-" + (id + 1))
      creator.addProperty(RDF.`type`, model.createResource(FOAF + "Agent"))
      creator.addProperty(RDF.`type`, model.createResource(PROV + "Agent"))
      dataid.addProperty(
        DC_11.creator,
        creator)
      creator.addProperty(
        model.createProperty(FOAF + "name"),
        cn)
      if(ce != "") {
        creator.addProperty(
          model.createProperty(FOAF + "mbox"),
          ce)
      }
    }

    for(((cn, ce), id) <- (CONTRIBUTOR_NAMES zip CONTRIBUTOR_EMAILS).zipWithIndex) {
      val creator = model.createResource(BASE_NAME + METADATA_PATH + "#Contributor-" + (id + 1))
      creator.addProperty(RDF.`type`, model.createResource(FOAF + "Agent"))
      creator.addProperty(RDF.`type`, model.createResource(PROV + "Agent"))
      dataid.addProperty(
        DC_11.creator,
        creator)
      creator.addProperty(
        model.createProperty(FOAF + "name"),
        cn)
      if(ce != "") {
        creator.addProperty(
          model.createProperty(FOAF + "mbox"),
          ce)
      }
    }

    for(de <- DERIVED_FROM) {
      dataid.addProperty(
        model.createProperty(PROV + "wasDerivedFrom"),
        model.createResource(de))
    }

    val dump = model.createResource(BASE_NAME + METADATA_PATH + "#Dump")

    dump.addProperty(RDF.`type`, model.createResource(DCAT + "Distribution"))

    dataid.addProperty(
      model.createProperty(DCAT + "distribution"),
      dump)

    dump.addProperty(
      model.createProperty(DCAT + "downloadURL"),
      model.createResource(BASE_NAME.dropRight(1) + DUMP_URI))

    dataid.addProperty( 
      model.createProperty(VOID + "triples"),
      backend.tripleCount.toString, NodeFactory.getType(XSD.integer.getURI()))

    dump.addProperty( 
      model.createProperty(VOID + "triples"),
      backend.tripleCount.toString, NodeFactory.getType(XSD.integer.getURI()))

    dump.addProperty(
      DC_11.format,
      "application/x-gzip")

    SPARQL_ENDPOINT match {
      case Some(se) =>
        dataid.addProperty(
          model.createProperty(VOID + "sparqlEndpoint"),
          model.createResource(se))
      case None =>
        dataid.addProperty(
          model.createProperty(VOID + "sparqlEndpoint"),
          model.createProperty(BASE_NAME.dropRight(1) + SPARQL_PATH))
    }

    for(((target, count), id) <- backend.linkCounts.zipWithIndex) {
      val linkset = model.createResource(BASE_NAME + METADATA_PATH + "#LinkSet-" + (id + 1))
      dataid.addProperty(
        model.createProperty(VOID + "subset"),
        linkset)
      linkset.addProperty(
        model.createProperty(VOID + "target"),
        model.createResource(target))
      linkset.addProperty(
        model.createProperty(VOID + "triples"),
        count.toString, NodeFactory.getType(XSD.integer.getURI()))
      linkset.addProperty(
        RDF.`type`,
        model.createResource(VOID + "LinkSet"))
    }
        
    model
  }
}

File: python/yuzu/dataid.py
from rdflib import Graph, Literal, URIRef
from rdflib.namespace import Namespace
from yuzu.backend import RDFBackend
from yuzu.settings import (DB_FILE, METADATA_PATH, DISPLAY_NAME, LANG,
                           BASE_NAME, LIST_PATH, ONTOLOGY, ISSUE_DATE,
                           VERSION_INFO, DESCRIPTION, LICENSE, KEYWORDS,
                           LICENSE_PATH, PUBLISHER_NAME, PUBLISHER_EMAIL,
                           CREATOR_EMAILS, CREATOR_NAMES, CONTRIBUTOR_EMAILS,
                           CONTRIBUTOR_NAMES, SPARQL_ENDPOINT, DERIVED_FROM,
                           DUMP_URI, SPARQL_PATH)
from rdflib.namespace import RDF, RDFS, XSD, DC, DCTERMS


def dataid():
    g = Graph()

    backend = RDFBackend(DB_FILE)

    BASE = Namespace(BASE_NAME)
    DCAT = Namespace("http://www.w3.org/ns/dcat#")
    VOID = Namespace("http://rdfs.org/ns/void#")
    DATAID = Namespace("http://dataid.dbpedia.org/ns#")
    FOAF = Namespace("http://xmlns.com/foaf/0.1/")
    ODRL = Namespace("http://www.w3.org/ns/odrl/2/")
    PROV = Namespace("http://www.w3.org/ns/prov#")

    dataid = BASE[METADATA_PATH]

    g.add((dataid, RDF.type, DCAT.Dataset))
    g.add((dataid, RDF.type, VOID.Dataset))

    g.add((dataid, DCAT.title, Literal(DISPLAY_NAME, lang=LANG)))

    g.add((dataid, RDFS.label, Literal(DISPLAY_NAME, lang=LANG)))

    g.add((dataid, DCAT.landingPage, URIRef(BASE_NAME)))

    g.add((dataid, VOID.exampleResource,
           URIRef(BASE[backend.list_resources(0, 1)[1][0]["link"][1:]])))

    g.add((dataid, DC.language, Literal(LANG)))

    g.add((dataid, VOID.rootResource, URIRef(BASE[LIST_PATH[1:]])))

    if ONTOLOGY:
        g.add((dataid, DATAID.ontologyLocation, URIRef(BASE[ONTOLOGY])))

    if ISSUE_DATE:
        g.add((dataid, DCTERMS.issued, Literal(ISSUE_DATE, datatype=XSD.date)))

    if VERSION_INFO:
        g.add((dataid, DATAID.versionInfo, Literal(VERSION_INFO)))

    if DESCRIPTION:
        g.add((dataid, DC.description, Literal(DESCRIPTION, lang=LANG)))

    g.add((dataid, ODRL.license, BASE[LICENSE_PATH[1:]]))

    if LICENSE:
        g.add((dataid, DC.rights, URIRef(LICENSE)))

    for keyword in KEYWORDS:
        g.add((dataid, DCAT.keyword, Literal(keyword)))

    if PUBLISHER_NAME:
        publisher = BASE[METADATA_PATH + "#Publisher"]
        g.add((publisher, RDF.type, FOAF.Agent))
        g.add((publisher, RDF.type, PROV.Agent))
        g.add((dataid, DC.publisher, publisher))
        g.add((publisher, FOAF.name, Literal(PUBLISHER_NAME)))
        g.add((publisher, FOAF.mbox, Literal(PUBLISHER_EMAIL)))

    for i in range(0, len(CREATOR_NAMES)):
        creator = BASE[METADATA_PATH + "#Creator-" + str(i + 1)]
        g.add((creator, RDF.type, FOAF.Agent))
        g.add((creator, RDF.type, PROV.Agent))
        g.add((dataid, DC.creator, creator))
        g.add((creator, FOAF.name, Literal(CREATOR_NAMES[i])))
        g.add((creator, FOAF.mbox, Literal(CREATOR_EMAILS[i])))

    for i in range(0, len(CONTRIBUTOR_NAMES)):
        creator = BASE[METADATA_PATH + "#Contributor-" + str(i + 1)]
        g.add((creator, RDF.type, FOAF.Agent))
        g.add((creator, RDF.type, PROV.Agent))
        g.add((dataid, DC.creator, creator))
        g.add((creator, FOAF.name, Literal(CONTRIBUTOR_NAMES[i])))
        g.add((creator, FOAF.mbox, Literal(CONTRIBUTOR_EMAILS[i])))

    for d in DERIVED_FROM:
        g.add((dataid, PROV.wasDerivedFrom, URIRef(d)))

    dump = BASE[METADATA_PATH + "#Dump"]

    g.add((dump, RDF.type, DCAT.Distribution))

    g.add((dataid, DCAT.distribution, dump))

    g.add((dump, DCAT.downloadURL, URIRef(BASE[DUMP_URI[1:]])))

    g.add((dataid, VOID.triples, Literal(str(backend.triple_count()),
                                         datatype=XSD.integer)))

    g.add((dump, VOID.triples, Literal(str(backend.triple_count()),
                                       datatype=XSD.integer)))

    g.add((dump, DC["format"], Literal("application/x-gzip")))

    if SPARQL_ENDPOINT:
        g.add((dataid, VOID.sparqlEndpoint, URIRef(SPARQL_ENDPOINT)))
    else:
        g.add((dataid, VOID.sparqlEndpoint, URIRef(BASE[SPARQL_PATH[1:]])))

    i = 0
    for target, count in backend.link_counts():
        linkset = BASE[METADATA_PATH + "#LinkSet-" + str(i + 1)]
        g.add((dataid, VOID.subset, linkset))
        g.add((linkset, VOID.subjectsTarget, dataid))
        g.add((linkset, VOID.target, URIRef(target)))
        g.add((linkset, VOID.triples, Literal(str(count),
               datatype=XSD.integer)))
        g.add((linkset, RDF.type, VOID.LinkSet))
        i += 1

    return g
